atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -2536,6 +2547,185 @@ public void exceptionThrownWhenTooManyPostFilterIterationsOccur() {
+    // with the empty byte array value.
+    @Test
+    public void storingMetadataIsAtomicWithLocalWrite() throws InterruptedException {
+        int numThreads = 20;
+        int numCells = 50;
+        long randomSeed = System.currentTimeMillis();
+
+        Random random = new Random(randomSeed);
+        SnapshotTransaction txn = unwrapSnapshotTransaction(txManager.createNewTransaction());
+
+        Set<Cell> cells = IntStream.range(0, numCells)
+                .mapToObj(cellIndex -> Cell.create(PtBytes.toBytes(cellIndex), PtBytes.toBytes(cellIndex)))
+                .collect(Collectors.toSet());
+
+        Map<Long, Map<Cell, ChangeMetadata>> metadataWrittenByThread =
+                concurrentlyWriteMetadataToCells(random, txn, cells, numThreads);
+
+        Map<Cell, ChangeMetadata> metadata = txn.getChangeMetadataForWrites(TABLE);
+        for (Cell cell : cells) {
+            byte[] currInternalValue = txn.get(TABLE, ImmutableSet.of(cell)).get(cell);
+            // The thread ID of the thread that won the race, i.e. the thread who was the last to write
+            long winnerThread = Longs.fromByteArray(currInternalValue);
+
+            ChangeMetadata expectedMetadata =
+                    metadataWrittenByThread.get(winnerThread).get(cell);
+            ChangeMetadata actualMetadata = metadata.get(cell);
+
+            assertThat(actualMetadata)
+                    .as(
+                            "Writing a value to the local buffer should be atomic with writing its metadata. If a"
+                                + " value was written without metadata, the metadata should be cleared. Random seed:"
+                                + " %d",
+                            randomSeed)
+                    .isEqualTo(expectedMetadata);
+        }
+    }
+
+    @Test
+    public void metadataIsSentToTimeLock() {
+        TimelockService timelockService = mock(TimelockService.class);
+        Transaction txn = getSnapshotTransactionWith(
+                timelockService, ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE_CELL)));
+
+        txn.putWithMetadata(
+                TABLE, ImmutableMap.of(TEST_CELL, ChangeMetadataAnnotatedValue.of(TEST_VALUE, TEST_METADATA)));
+        maybeCommit(txn);
+
+        verifyLockWasCalledWithMetadata(
+                timelockService,
+                Optional.of(LockRequestMetadata.of(ImmutableMap.of(
+                        AtlasCellLockDescriptor.of(
+                                TABLE.getQualifiedName(), TEST_CELL.getRowName(), TEST_CELL.getColumnName()),
+                        TEST_METADATA))));
+    }
+
+    @Test
+    public void noMetadataResultsInEmptyOptional() {
+        TimelockService timelockService = mock(TimelockService.class);
+        Transaction txn = getSnapshotTransactionWith(timelockService, ImmutableMap.of());
+        maybeCommit(txn);
+        verifyLockWasCalledWithMetadata(timelockService, Optional.empty());
+    }
```

Comment:
Although this is not a necessary invariant, we really, really like to have this for performance reasons. What if we have a transaction that does not use metadata?. We would like to avoid having to do any index encoding.